File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        self.cyclicsort(nums)

        for i in range (len(nums)):
            if(nums[i] != i + 1):
                return nums[i]
            
        

        return -1
    

    def cyclicsort(self, nums: List[int]):
        i = 0
        while(i < len(nums)):
            correct = nums[i] - 1

            if(nums[i] != nums[correct]):
                self.swap(nums, i, correct)
            else :
                i+=1
            
        
    

    def swap(self, nums: List[int],  i, j):
        temp = nums[i]
        nums[i] = nums[j]
        nums[j] = temp
    

File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        cyclicsort(nums);

        for(int i = 0; i < nums.length; i++){
            if(nums[i] != i + 1){
                return nums[i];
            }
        }

        return -1;
    }

    void cyclicsort(int[] nums){
        int i = 0;
        while(i < nums.length){
            int correct = nums[i] - 1;

            if(nums[i] != nums[correct]){
                swap(nums, i, correct);
            } else {
                i++;
            }
        }
    }

    void swap(int[] nums, int i, int j){
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}